File: python/Baekjoon/9205.py
import sys
from collections import deque

T = int(sys.stdin.readline().strip())
result = []

for t in range(T):
    martsCount = int(sys.stdin.readline().strip())
    home = list(map(int, sys.stdin.readline().strip().split()))
    marts = []
    for _ in range(martsCount):
        marts.append(list(map(int, sys.stdin.readline().strip().split())))
    festival = list(map(int, sys.stdin.readline().strip().split()))

    def bfs():
        queue = deque([home])
        visited = [False for _ in range(martsCount)]

        while queue:
            x, y = queue.popleft()

            if abs(x - festival[0]) + abs(y - festival[1]) <= 50 * 20:
                result.append("happy")
                return

            for i in range(martsCount):
                if not visited[i]:
                    if abs(x - marts[i][0]) + abs(y - marts[i][1]) <= 50 * 20:
                        queue.append((marts[i][0], marts[i][1]))
                        visited[i] = True

        result.append("sad")

    bfs()
    
print("\n".join(result))

File: Baekjoon/9205.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const [T, ...input] = fs.readFileSync(filePath).toString().trim().split("\n");
const result = [];
let idx = 0;

for (let t = 0; t < T; t++) {
  const martsCount = input[idx++].split(" ").map(Number);
  const home = input[idx++].split(" ").map(Number);
  const marts = [];
  for (let m = 0; m < martsCount; m++) {
    marts.push(input[idx++].split(" ").map(Number));
  }
  const festival = input[idx++].split(" ").map(Number);

  const visited = Array.from({ length: martsCount }, () => false);
  const queue = [];
  let head = 0;

  const bfs = () => {
    queue.push(home);

    while (queue.length > head) {
      const [curX, curY] = queue[head++];

      if (Math.abs(festival[0] - curX) + Math.abs(festival[1] - curY) <= 1000)
        return result.push("happy");

      for (let i = 0; i < martsCount; i++) {
        if (
          !visited[i] &&
          Math.abs(marts[i][0] - curX) + Math.abs(marts[i][1] - curY) <= 1000
        ) {
          queue.push(marts[i]);
          visited[i] = true;
        }
      }
    }
    return result.push("sad");
  };

  bfs();
}

return console.log(result.join("\n"));
